anvil_consortium_manager/models.py

```diff
@@ -72,17 +75,125 @@ def anvil_import(cls, billing_project_name):
         return billing_project
 
 
+class UserEmailEntry(TimeStampedModel, models.Model):
+    """A model to store emails that users could link to their AnVIL account after verification."""
+
+    uuid = models.UUIDField(default=uuid.uuid4)
+    """UUID for use in urls."""
+
+    email = models.EmailField()
+    """The email entered by the user."""
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, null=True, on_delete=models.CASCADE
+    )
+    """The user who created the record."""
+
+    date_verification_email_sent = models.DateTimeField()
+    """Most recent date that a verification email was sent."""
+
+    date_verified = models.DateTimeField(null=True, blank=True)
+    """The date that the email was verified by the user."""
+
+    history = HistoricalRecords()
+    """Django simple history."""
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=["email", "user"], name="unique_user_email_entry"
+            )
+        ]
+
+    def __str__(self):
+        """String method."""
+        return "{email} for {user}".format(email=self.email, user=self.user)
+
+    def anvil_account_exists(self):
+        """Check if this account exists on AnVIL."""
+        try:
+            AnVILAPIClient().get_proxy_group(self.email)
+        except AnVILAPIError404:
+            return False
+        return True
+
+    def save(self, *args, **kwargs):
+        self.email = self.email.lower()
+        return super().save(*args, **kwargs)
+
+
 class Account(TimeStampedModel, ActivatorModel):
     """A model to store information about AnVIL accounts."""
 
+    ERROR_USER_WITHOUT_VERIFIED_EMAIL_ENTRY = (
+        "Accounts with a user must have a verified_email_entry."
+    )
+    ERROR_VERIFIED_EMAIL_ENTRY_WITHOUT_USER = (
+        "Accounts with a verified_email_entry must have a user."
+    )
+    ERROR_UNVERIFIED_VERIFIED_EMAIL_ENTRY = (
+        "verified_email_entry must have date_verified."
+    )
+    ERROR_MISMATCHED_USER = "Account.user and verified_email_entry.user do not match."
+    ERROR_MISMATCHED_EMAIL = (
+        "Account.email and verified_email_entry.email do not match."
+    )
+
     # TODO: Consider using CIEmailField if using postgres.
     email = models.EmailField(unique=True)
+    # Use on_delete=PROTECT here because additional things need to happen when an account is deleted,
+    # and we're not sure what that should be. Deactivate the account or and/or remove it from groups?
+    # I think it's unlikely we will be deleting users frequently, and we can revisit this if necessary.
+    # So table it for later.
+    user = models.OneToOneField(
+        settings.AUTH_USER_MODEL, null=True, blank=True, on_delete=models.PROTECT
+    )
```

Comment:
Should we store this here, or just access it via `account.verified_email_entry.user`? It is less duplication of information, but it means that an account linked to a user must have a `verified_email_entry`. This is already required by the clean method, so I think it's ok.

Reply:
Leaning towards no, because then we can't enforce the one user -> one Account as easily.